src/core/analog_controller.cpp

```diff
@@ -271,350 +262,370 @@ void AnalogController::SetMotorStateForConfigIndex(int index, u8 value)
 
-      FIXED_REPLY_STATE(State::Command4C1, 0x00, true, State::Command4C2);
-      FIXED_REPLY_STATE(State::Command4C2, 0x00, true, State::Command4C3);
+        if (m_rumble_config_small_motor_index == -1)
+          SetMotorState(SmallMotor, 0);
 
-    case State::Command4C3:
-    {
-      // Ape Escape sends both 0x00 and 0x01 sequences on startup and checks for correct response
-      if (m_command_param == 0x00)
-        *data_out = 0x04;
-      else if (m_command_param == 0x01)
-        *data_out = 0x07;
-      else
-        *data_out = 0x00;
+        if (m_rumble_config_large_motor_index == -1 && m_rumble_config_small_motor_index == -1)
+          m_rumble_unlocked = false;
+      }
 
-      m_state = State::Command4C4;
-      ack = true;
+      // Unknown if motor config array forces 0xFF values if configured byte is not 0x00 or 0x01
+      // Unknown under what circumstances rumble is locked and legacy rumble is re-enabled, if even possible
+      // Current assumption is that rumble is locked and legacy rumble is re-enabled when new config is all 0xFF
     }
     break;
 
-      FIXED_REPLY_STATE(State::Command4C4, 0x00, true, State::Command4C5);
-      FIXED_REPLY_STATE(State::Command4C5, 0x00, false, State::Idle);
-
-      ID_STATE_MSB(State::UnlockRumbleIDMSB, State::GetSetRumble1);
-      REPLY_RUMBLE_CONFIG(State::GetSetRumble1, 0, true, State::GetSetRumble2);
-      REPLY_RUMBLE_CONFIG(State::GetSetRumble2, 1, true, State::GetSetRumble3);
-      REPLY_RUMBLE_CONFIG(State::GetSetRumble3, 2, true, State::GetSetRumble4);
-      REPLY_RUMBLE_CONFIG(State::GetSetRumble4, 3, true, State::GetSetRumble5);
-      REPLY_RUMBLE_CONFIG(State::GetSetRumble5, 4, true, State::GetSetRumble6);
-
-    case State::GetSetRumble6:
-    {
-      DebugAssert(5 < m_rumble_config.size());
-      *data_out = m_rumble_config[5];
-      m_rumble_config[5] = data_in;
-
-      if (data_in == 0x00)
-        m_rumble_config_small_motor_index = 5;
-      else if (data_in == 0x01)
-        m_rumble_config_large_motor_index = 5;
-
-      if (m_rumble_config_large_motor_index == -1)
-        SetMotorState(LargeMotor, 0);
+      DefaultCaseIsUnreachable();
+  }
 
-      if (m_rumble_config_small_motor_index == -1)
-        SetMotorState(SmallMotor, 0);
+  *data_out = m_tx_buffer[m_command_step];
 
-      if (m_rumble_config_large_motor_index == -1 && m_rumble_config_small_motor_index == -1)
-        m_rumble_unlocked = false;
+  m_command_step = (m_command_step + 1) % m_response_length;
+  ack = (m_command_step == 0) ? false : true;
 
-      // Unknown if motor config array forces 0xFF values if configured byte is not 0x00 or 0x01
-      // Unknown under what circumstances rumble is locked and legacy rumble is re-enabled, if even possible
-      // e.g. need all 0xFFs?
+  if (m_command_step == 0)
+  {
+    m_command = Command::Idle;
 
-      m_state = State::Idle;
-      ack = false;
-    }
-    break;
+    std::string tmp_str = "";
```

Comment:
Not really a fan of the heap allocations here - can we put this behind a `#ifdef _DEBUG`? (or use `SmallString`, which is stack-allocated).

But otherwise a nice cleanup!

Reply:
Ah shoot -- meant to get rid of the heap allocation (since it's clear anyways what the transfer lengths are), but forgot to do so. Thanks for the catch!